Trigger Migration Flow when launching the app from deep links

ovitrif opened the pull request:
Fixes #17653

This PR implements the logic to:
- Start the migration flow when the app is launched from deep links
- Land on the screen from the deep link after migration

> **Note** **Known issue**: When the Jetpack migration flow is started from a deep link, the notification permissions dialog is not shown on first launch of a freshly installed app. When the app is force closed and opened the second time, the dialog to enable notifications appears. Since the same behavior was already reproducible in the WordPress app this is not a blocker imho.

## To Test

#### Prerequisites & Setup

- ⚠️ Uninstall all versions of the WordPress and Jetpack apps from the device (including pre-alphas).
- ↗️ Open the [**links helper webpage**](https://codepen.io/ovitrif/full/WNKRBeV) in the device web browser & keep the browser open on this page.
- Clearing app data is required multiple times during testing, see **🆕 How to clear app data?** panel for steps.
- Setting which app to handle links is also needed, see **☑️ How to set app to handle verified links?** panel.

<details>
<summary><b>🆕 How to clear app data?</b></summary>

1. Open the apps drawer on the device
2. Tap & hold the icon of the desired app
3. Select `App Info` 
4. Tap `Storage` / `Storage & cache` (whichever presented)
5. Tap `Clear data` / `Clear storage` (whichever presented)
---
</details>

<details>
<summary><b>☑️ How to set app to handle verified links?</b></summary>

1. Open the apps drawer on the device
2. Tap & hold the icon of the desired app
3. Select `App Info` 
4. Tap `Open by default`
5. Tap `+ Add link`
6. ☑️ Check all three checkboxes then tap `Add`
---
</details>

#### 1. Jetpack App: Complete migration flow from `web link`

1. Build & Install WordPress app and log in (flavor: `wordpressWasabiDebug`)
2. Install Jetpack app of the same flavor & close it, then 🆕 clear app data
3. ☑️ Set Jetpack app to handle verified links
4. ↗️ Open `Stats` web link
5. **Expect** to see migration flow
7. Proceed to finish migration
8. **Expect** to land on the `Stats` screen
9. **Verify** your data was migrated

#### 2. Jetpack App: Complete migration flow from `deep link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ↗️ Open `Jetpack Reader` deep link with Jetpack app (just once)
4. **Expect** to see migration flow
5. Proceed to finish migration
6. **Expect** to land on the `Reader` screen

#### 3. Jetpack App: Complete migration flow from `post link`

1. **Prerequisite** WordPress app installed and logged in  (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links
4. ↗️ Open `Post` web link
5. **Expect** to see migration flow
6. Proceed to finish migration
7. **Expect** to land on `Reader` screen with the post `Post deep link test` open

#### 4. Jetpack App: Interrupt and resume migration from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen change device settings to `dark mode`
7. ~~Put the Jetpack app in background and bring it back to foreground~~ ([skip this step](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issuecomment-1381980835))
9. **Expect** The migration flow to continue
10. On the migration Welcome screen tap `continue`
11. Disable `dark theme`
12. **Expect** to remain on the migration Notifications screen
13. Proceed to finish migration
14. **Expect** To land on `Notifications` screen

#### 5. Jetpack App: Force close migration from `web link` and restart

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see migration flow
6. Force close the Jetpack app
7. Reopen the Jetpack app
8. **Expect** to see migration flow
10. Proceed to finish migration
11. **Expect** to land on `My Site` screen
   **Note:** Even though it could be an option to land on the screen from the web link (`Stats`), this wouldn't be ideal because users could open the Jetpack app after a long time.

#### 6. Jetpack App: Exit migration and login from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen navigate back
7. **Expect** to see login screen
8. Login
9. **Expect** to land on the `My Site` screen
10. **Verify** that your data is not migrated (if you used the same account from WP app)

#### 7. Jetpack App: Open from `web link` without migration

1. 🆕 Clear WordPress app data
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see login screen in Jetpack app
6. Login
7. **Expect** to land on the `Stats` screen

#### 8. WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

1. 🆕 Clear WordPress app data (should be the case)
2. Open WordPress app and login
3. 🆕 Clear Jetpack app data
4. ☑️ Set **WordPress** app to handle verified links
5. ↗️ Open `Notifications` web link
6. **Expect** to see `Open links in Jetpack?` screen
7. ☑️ Set **Jetpack** app to handle verified links
8. Return to WordPress app
9. Tap `Open links in Jetpack`
10. **Expect** to see migration flow
15. Proceed to finish migration
16. **Expect** to land on `Notifications` screen

#### 9. Regression: Jetpack App migration not eligible with Local Draft Content

1. **Prerequisite** WordPress app installed and logged in
2. Open the WordPress app and go to `Posts`
3. Disable internet connection on the device
4. Create a new blog post, add title and body text
5. Navigate `back` from the editor to save the post as local draft
6. Clear Jetpack app data
8. Open the Jetpack app
9. **Expect** to land on the login/ landing screen
10. Restore internet connection on the device
13. **Optionally** login and verify that login is successful

Eventually think of other regression testing scenarios that could occur during the migration flow and when opening the app from deep links and test them.

## Regression Notes
1. Potential unintended areas of impact
   Migration flow & deep link flows.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing & relied on existing migration and deep links tests.

3. What automated tests I added (or what prevented me from doing so)
   Added test for the updated logic in `JetpackAppMigrationFlowUtilsTest`.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ovitrif committed:
Start migration flow when JP is opened from deep link

ovitrif committed:
Resume deep link handling after completing migration

When the app is opened from a deep link, the deep link data is passed through the migration flow so at the end of the flow we can navigate to the screen specified in the deep link.

ovitrif committed:
Merge branch 'try/reset-migration-state-when-interrupted' into issue/17653-jp-migration-web-links

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/main/WPMainActivity.java
#	WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationFragment.kt
#	WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationViewModel.kt

ovitrif committed:
Refactor deep link migration to use only data class

Simplified the logic to handle deep link opening after the migration flow is completed by relying only on the deep link data class

ovitrif committed:
Refactor to keep complete flows tidy

ovitrif committed:
Update comment for migration deep link entry point

ovitrif committed:
Don't finish deep linking activity when migration starts

This fix was required to make sure the notifications permission dialog is presented when launching the migration flow from a deep link, similar to the default behavior when the app is normally started.

ovitrif committed:
Open deep link after migration only if it has valid data

ovitrif committed:
Pre-check migration eligibility before starting the flow

ovitrif committed:
Add overload for starting migration flow without deep link

ovitrif committed:
Merge branch 'trunk' into issue/17653-jp-migration-web-links

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationActivity.kt
#	WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationFragment.kt
#	WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationViewModel.kt

ovitrif committed:
Fix formatting issues in viewmodel

ovitrif committed:
Move eligibility check after compatibility check

This change ensures the WordPress app compatibility is tracked

ovitrif committed:
Adapt migration flow unit tests to new logic and add test

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/ActivityLauncher.java:
```diff
@@ -1775,9 +1778,21 @@ public static void showLoginPrologue(@NonNull Context context) {
+                                                 @Nullable PreMigrationDeepLinkData deepLinkData) {
         Intent intent = new Intent(context, JetpackMigrationActivity.class);
         intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
+        if (deepLinkData != null) {
+            intent.putExtra(KEY_DEEP_LINK_DATA, deepLinkData);
+        }
+        context.startActivity(intent);
+    }
+
+    public static void openDeepLinkAfterJPMigration(@NonNull Context context, String action, Uri uri) {
+        Intent intent = new Intent(context, DeepLinkingIntentReceiverActivity.class);
+        intent.setAction(action);
+        intent.setData(uri);
+        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);
         context.startActivity(intent);
```
@zwarm I added you as reviewer (since you're the author of the latest deep links changes) to please verify if you don't foresee any possible issues with running this code at the end of the migration flow in order to open the app at the screen targeted in the deep links 🙏 🙇‍♂️

ovitrif replied:
The code was changed after the implementation to trigger migration from post web links (https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/a312e88132621713dc6012362e956a59b86a4c27), this is no longer a concern because the Jetpack app is [set explicitly to handle the link after migration](https://github.com/wordpress-mobile/WordPress-Android/pull/17742/files#diff-200f30b456241b2103300badbb3b04bfe696ce8386f45a8781c7d95db564c677R1790-R1795).

zwarm commented:
@ovitrif - I am stuck on the first test because neither app has been verified to handle web links, so the link opens with the browser. I can manually assign Jetpack or WP as the default handler for verified web links, but not both. I found the best way to test the seamless redirect scenarios was to create release versions of each app, not sure that is an option here. 

In addition I didn't see a link in your test html for the following type of links. 
Shared link from the post itself and/or from comment/like emails (including clicking on titles within the email and blog names). For instance: `https://zieglerkirby.wordpress.com/2022/10/18/annmarie/` 

Coming into WP from a link like this will launch the Open Links with Jetpack overlay. `https://zieglerkirby.wordpress.com/2022/10/18/annmarie/`

ovitrif commented:
> @ovitrif - I am stuck on the first test because neither app has been verified to handle web links, so the link opens with the browser.

Thank you for your remarks @zwarm 👍 

While testing using Android Studio debug builds locally I could open the links in the app if **using a physical device**. This didn't work in the emulator for me as well.

I also tested now with the pre-alpha builds ([WordPress](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issuecomment-1378926481) & [Jetpack](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issuecomment-1378925731) from the PR.

I was thinking if perhaps the issue you're encountering is because the Browser is set to open such links by default on your device, but it seems to still be possible to open the links with the apps on my device even if that's enabled:

<details>
<summary> ↘️ Preview when Chrome is set to open such links by default</summary>

<img width="314" src="https://user-images.githubusercontent.com/4588074/211885577-38943e08-5f24-4fda-ada4-53bcc05287e8.png">
</details>


> In addition I didn't see a link in your test html for the following type of links.
> - Shared link from the post itself

Thank you for informing me to add such links 🙇 , I'll add this tomorrow to the testing HTML and add/change a test to use such links 👍.

zwarm commented:
@ovitrif 

> While testing using Android Studio debug builds locally I could open the links in the app if using a physical device. This didn't work in the emulator for me as well.

I am using a physical device (Pixel 5 running Android 13.). I had to go into the app setting for WP and set the verified links manually, but then JP doesn't open them by default. Which app should be the default handler for the above tests? When using release builds, both apps will handle the links; however it the most recently installed app that will actually launch.

> I was thinking if perhaps the issue you're encountering is because the Browser is set to open such links by default on your device, but it seems to still be possible to open the links with the apps on my device even if that's enabled:

I am not prompted with the disambiguation dialog; I go straight to the browser. 

FYI: The link I shared has a different receiver and is handled in`ReaderPostPagerActivity`

Here are the steps I took for the 1st test
    - Install WP + log in
    - Install JP then close it and clear app data
    - Click on stats link - taken to browser not JP
    - Navigate to app settings for JP and enable verified links
    - Click the stats link
    - Jetpack is launched on the login screen. 
    - Logged in
    - Presented with the sites list, but no site selection available, just the “done” button
    - Tapped “done”
    - Taken to the stats view
    - Did not see the migration flow

ovitrif commented:
> FYI: The link I shared has a different receiver and is handled in`ReaderPostPagerActivity`

Thank you for the additional info! That means there should be logic to trigger the migration also from `ReaderPostPagerActivity` 👍 
I'll work on adapting the code to handle it 👍 and see what other changes may be needed, as well as adding a test scenario for post links.

> Here are the steps I took for the 1st test
> [...]

Thanks, I'll test this flow and share my findings 👍 

> Which app should be the default handler for the above tests?
@zwarm Regarding the default handler, the steps for opening the app from a link have this format:

> #### 1. Complete migration flow from web link
> ...
> 3. ↗️ Open Stats web link with **Jetpack** app (just once)

The handler should be the app that is mentioned, in the case of the first test, **Jetpack** app.

In any case if this proves to be an issue for testing, I'll adapt the steps according to what would be the better option:
- If using **release** builds, then adapt instructions to use release builds
- If manually **enable verified links** for the handler app, then add steps for this

ovitrif commented:
> Here are the steps I took for the 1st test
> - Install WP + log in
> - Install JP then close it and clear app data
> - Click on stats link > taken to browser not JP
> - Navigate to app settings for JP and enable verified links
> - Click the stats link
> - Jetpack is launched on the login screen.
> - Logged in
> - Presented with the sites list, but no site selection available, just the “done” button
> - Tapped “done”
> - Taken to the stats view
> - Did not see the migration flow

> Thanks, I'll test this flow and share my findings 👍

@zwarm In my tests following this steps opened the migration flow 🤷 .
I'm not sure what could be the issue but FYI the migration eligibility is checked before showing the migration UI, and if it fails for any reason then it's not shown. It might have been a different local setup 🤔 causing the migration to be ineligible, I can't think of any other reason.

Anyways I've updated the testing instructions with steps to set which app to handle verified links and went through all testing scenarios with the `wasabiDebug` variant of the apps, after the test cases adjustments they all succeeded for me ✅.

ovitrif committed:
Trigger migration flow for post links

ovitrif commented:
PR updated to open post links as well: https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/a312e88132621713dc6012362e956a59b86a4c27, I a test case to cover post links:

> #### [3. Jetpack App: Complete migration flow from post link](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issue-1529210981)

With the latest updates I couldn't find any issue by following the test cases both on emulator or physical device.

The PR looks ready on my side 👋

ovitrif committed:
Merge branch 'trunk' into issue/17653-jp-migration-web-links

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/utils/JetpackAppMigrationFlowUtils.kt:
```diff
@@ -30,11 +36,18 @@ class JetpackAppMigrationFlowUtils @Inject constructor(
             && !appPrefsWrapper.isJetpackMigrationCompleted()
             && isWordPressInstalled()
             && isWordPressCompatible()
+            && isMigrationEligible()
```
Is it necessary to invoke the eligibility check here?

The main concern I have about this is that `shouldShowMigrationFlow` is usually called from the UI thread, and this has the potential be somewhat IO-heavy. This is usually part of the migration itself, since behind the scenes, it is using a content resolver to [query all post and page content on every site](https://github.com/wordpress-mobile/WordPress-Android/blob/a01543b9d72732e07fe2b43bf479ba90fb2a24cd/WordPress/src/main/java/org/wordpress/android/localcontentmigration/LocalEligibilityStatusProviderHelper.kt#L26) in the WordPress app.

We have the `isJetpackMigrationEligible()` flag in the prefs, but I imagine that this was added here because this is needed for the first go around (let me know if that is a wrong assumption :sweat_smile: ). To handle that scenario, I wonder if we could somehow utilize the `Ineligible` state transition [in the viewModel](https://github.com/wordpress-mobile/WordPress-Android/blob/536e24f4c2f87ad4c0b0e7007504815be96549d0/WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationViewModel.kt#L95)? Wdyt?

ovitrif replied:
> Is it necessary to invoke the eligibility check here?

Good question! 👍 

> We have the `isJetpackMigrationEligible()` flag in the prefs, but I imagine that this was added here because this is needed for the first go around (let me know if that is a wrong assumption 😅 ).

It was indeed the case, without it the flow via deep link was also not working correctly in situations where:
1. The WordPress app is not logged in → JP Migration error `WPNotLoggedIn`
  In this case the desired functionality of opening the Jetpack app from links should be the same as without this PR, the Login screen should be shown and the user would end up on the correct page after the login.
2. There are local draft changes in the WordPress app not synced with the backend → Error `LocalDraftContentIsPresent`.

> To handle that scenario, I wonder if we could somehow utilize the Ineligible state transition [in the viewModel](https://github.com/wordpress-mobile/WordPress-Android/blob/536e24f4c2f87ad4c0b0e7007504815be96549d0/WordPress/src/main/java/org/wordpress/android/ui/main/jetpack/migration/JetpackMigrationViewModel.kt#L95)? Wdyt?

I already explored this solution to some extent, but it wasn't producing the desired result (see point 1 above). I'm not sure anymore what else wasn't ok with it, but I can look more into this suggested approach 👍

mkevins commented:
Hey Ovi, this is working very well for me so far! :+1: 

Some things that I noticed when testing:

* There were 4 links (in the link handling settings) that I could enable instead of 3, so I just enabled all of them.

:heavy_check_mark:  1. Jetpack App: Complete migration flow from `web link`

:heavy_check_mark:  2. Jetpack App: Complete migration flow from `deep link`

:heavy_check_mark:  3. Jetpack App: Complete migration flow from `post link`

:question:  4. Jetpack App: Interrupt and resume migration from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen change device settings to `dark mode`
7. Put the Jetpack app in background and bring it back to foreground 
    * :question:  When I did this, the Jetpack app was gone - seemingly it was still part of the browser's backstack, so maybe this is a browser behavior?
9. **Expect** The migration flow to continue
10. On the migration Welcome screen tap `continue`
11. Disable `dark theme`
12. **Expect** to remain on the migration Notifications screen
13. Proceed to finish migration
14. **Expect** To land on `Notifications` screen

Everything else about this test worked as expected.

:heavy_check_mark:  5. Jetpack App: Force close migration from `web link` and restart

:heavy_check_mark:  6. Jetpack App: Exit migration and login from `web link`

:heavy_check_mark:  7. Jetpack App: Open from `web link` without migration

1. 🆕 Clear WordPress app data
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see login screen in Jetpack app
6. Login
    * Note: first I saw the site selection screen, but was eventually brought to the stats screen

:heavy_check_mark:  8. **Expect** to land on the `Stats` screen

8 was a tricky one because I didn't realize at first that I should manually set the link handler. After re-reading, though, and considering that these are not signed releases, this makes sense. After performing the steps correctly, it works as expected.

:heavy_check_mark:  9. WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

mkevins commented:
There is one step in the migration that has not been standardized / unified yet (but I believe it is ready to be unified with the rest of the steps, since we now preserve site ids). More details in [this issue](https://github.com/wordpress-mobile/WordPress-Android/issues/17752) which I've just opened. This _might_ be relevant (a potential pre-requisite?) to this PR if any of the destination links after migration result in the blogging reminders not being migrated. I plan to work on this early next week. I'm still unsure whether this is a blocker for this PR, but the broken scenario I had in mind is:

1. Migrate in any of the test steps above with a destination link that does not also load `WPMainActivity` before the `onSiteChanged` event occurs.
2. Leave the app
3. Delete the WordPress app.
4. :boom:  Blogging Reminders are lost

I have not tested or confirmed this in practice, but it came to mind while testing these flows.

ovitrif commented:
> There is one step in the migration that has not been standardized / unified yet (but I believe it is ready to be unified with the rest of the steps, since we now preserve site ids). More details in [this issue](https://github.com/wordpress-mobile/WordPress-Android/issues/17752) which I've just opened. This _might_ be relevant (a potential pre-requisite?) to this PR if any of the destination links after migration result in the blogging reminders not being migrated. I plan to work on this early next week. I'm still unsure whether this is a blocker for this PR, but the broken scenario I had in mind is:
> 
> 1. Migrate in any of the test steps above with a destination link that does not also load `WPMainActivity` before the `onSiteChanged` event occurs.
> 2. Leave the app
> 3. Delete the WordPress app.
> 4. 💥  Blogging Reminders are lost
> 
> I have not tested or confirmed this in practice, but it came to mind while testing these flows.

Great thinking and very nice to think of this scenario, thank you for considering it and filing the [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/17752) @mkevins 👍 

I'm not sure myself if it's a blocker since before this PR the migration wouldn't even trigger, while after it does but it may result in blogging reminders data loss if it's one of the affected cases (destination link that does not also load `WPMainActivity` before the `onSiteChanged` event occurs).
So it's either all data from WordPress app lost if users remain in the Jetpack app or just blogging reminders lost 🤷

ovitrif commented:
@mkevins Thanks a lot for testing all test cases 🙇 

Regarding 4th test case:
<details>
<summary>(Click for full steps) <h4>4. Jetpack App: Interrupt and resume migration from <code>web link</code></h4></summary>

> 
> 1. **Prerequisite**: WordPress app installed and logged in (should be the case)
> 2. 🆕 Clear Jetpack app data
> 3. ☑️ Set Jetpack app to handle verified links (should be the case)
> 4. ↗️ Open `Notifications` web link
> 5. **Expect** to see migration flow
> 6. On the migration screen change device settings to `dark mode`
> 7. Put the Jetpack app in background and bring it back to foreground
>    * ❓  When I did this, the Jetpack app was gone - seemingly it was still part of the browser's backstack, so maybe this is a browser behavior?
> 8. **Expect** The migration flow to continue
> 9. On the migration Welcome screen tap `continue`
> 10. Disable `dark theme`
> 11. **Expect** to remain on the migration Notifications screen
> 12. Proceed to finish migration
> 13. **Expect** To land on `Notifications` screen

---
</details>

> ❓ 4. Jetpack App: Interrupt and resume migration from `web link`
> 7. Put the Jetpack app in background and bring it back to foreground
>    * ❓  When I did this, the Jetpack app was gone - seemingly it was still part of the browser's backstack, so maybe this is a browser behavior?

I encountered the same issue with any link type. When looking for what might cause it I stumbled upon the [`android:excludeFromRecents="true"`](https://github.com/wordpress-mobile/WordPress-Android/blob/3b3b7d8e1d1b2e5d3a0bc4985da978de271e2420/WordPress/src/main/AndroidManifest.xml#L392) and figured out this might be a side effect of an actually intended behavior 😅 . The idea is that users would go back to the browser window when they tap the back button, which looks valid to me 👍 , and this is the same behavior in production or on trunk.

I guess for testing purposes I could remove the step about putting the app in background 🤷 , Thanks for pointing it out though 👋

mkevins commented:
> So it's either all data from WordPress app lost if users remain in the Jetpack app or just blogging reminders lost

Right, I agree that this is a step in the right direction, but I think in that first case (where nothing is migrated) we aren't suggesting to delete the WordPress app - which is really the main impetus for caution, imo (i.e. since we are suggesting for the user to perform a potentially irreversible destructive / lossy action, we have to be extra careful that all data is preserved). In any case, I'll work on this today to hopefully make it a moot point :smile: .

ovitrif commented:
> but I think in that first case (where nothing is migrated) we aren't suggesting to delete the WordPress app - which is really the main impetus for caution, imo (i.e. since we are suggesting for the user to perform a potentially irreversible destructive / lossy action, we have to be extra careful that all data is preserved).

Good point, indeed when nothing is migrated we aren't suggesting to delete the WordPress app 👍 

> In any case, I'll work on this today to hopefully make it a moot point 😄 .

Thank you for taking the initiative to improve this flow 🙇

ovitrif committed:
Clear migrationInProgress flag after manual login

This change was needed to fix the issues where

ovitrif committed:
Use bool primitive for shouldBypassMigration

ovitrif commented:
### Update
- Merged the changes from the [wip branch](https://github.com/wordpress-mobile/WordPress-Android/compare/issue/17653-jp-migration-web-links...issue/17653-jp-migration-web-links--suggestion-ineligible-wip) of @mkevins 
-  added an additional fix in https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/db745bd3bc92c0c1b9e35b24a3ac66da825b5f07 to solve the issue with the app remaining logged out after login from deep link
- cleaned up imports (https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/65e6ff50f883c792496904f70a1db96b718ba13a) and tests https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/0ceead305039c37aee389edf0ee962f6478e683e

This is ready for another pass 😉

ovitrif committed:
Remove unused imports

ovitrif committed:
Remove eligibility checks in JetpackAppMigrationFlowUtilsTest

ovitrif commented:
### Update

As discussed with @mkevins although the PR seems ready, the plan is to:
- Merge after #17778

This way we avoid introducing unexpected cases where users could migrate to the Jetpack App via a `Reader` deep link ending up without their blogging reminder(s) migrated from the WordPress app.

mkevins committed:
Add migration bypass argument to deeplink activity launcher

mkevins committed:
Enable bypassing migration in deep link intent receiver

mkevins committed:
Add optional deep link data to completion and fallback action events

mkevins committed:
Refactor action event handling for optional deep link destinations

mkevins committed:
Remove eligibility helper from migration flow utils

mkevins committed:
Set migration to ineligible after a successful login

Migration can be re-enabled after logging out by either deleting the
app, or clearing app data.

mkevins committed:
Merge branch 'trunk' into issue/17653-jp-migration-web-links--suggestion-rebase

mkevins commented:
Thanks for these updates Ovi :+1: .

Since conflicts have arisen, I created [a new branch with a rebase of this branch](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17653-jp-migration-web-links--suggestion-rebase) (which resolves the conflicts, and also splits out the `wip` commit from my earlier suggestion branch). However, after resolving the conflicts and retesting the above scenarios, I found that some of them are no longer working as expected :disappointed: . I haven't had a chance to dig into why yet, so hoping it is something simple :crossed_fingers: .

<details>
<summary>
❌ These tests were performed on a Pixel 3a (physical device) with preconditions
</summary>


WordPress app settings were changed:
* Dark mode (OS default is light mode)
* Language Spanish (OS default is English)

#### 1. ✅ Jetpack App: Complete migration flow from `web link`

#### 2. ❌ Jetpack App: Complete migration flow from `deep link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ↗️ Open `Jetpack Reader` deep link with Jetpack app (just once)
4. **Expect** to see migration flow
5. Proceed to finish migration
6. ❌ **Expect** to land on the `Reader` screen

I landed on the my site screen.

#### 3. ✅ Jetpack App: Complete migration flow from `post link`

#### 4. ❌ Jetpack App: Interrupt and resume migration from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen change device settings to `dark mode`
7. ~~Put the Jetpack app in background and bring it back to foreground~~ ([skip this step](https://github.com/wordpress-mobile/WordPress-Android/pull/17742#issuecomment-1381980835))
9. **Expect** The migration flow to continue
10. On the migration Welcome screen tap `continue`
11. Disable `dark theme`
12. **Expect** to remain on the migration Notifications screen
13. Proceed to finish migration
14. ❌ **Expect** To land on `Notifications` screen

I landed on the my site screen.

#### 5. ❌ Jetpack App: Force close migration from `web link` and restart

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Stats` web link
5. **Expect** to see migration flow
6. Force close the Jetpack app
7. Reopen the Jetpack app
8. ❌ **Expect** to see migration flow
10. Proceed to finish migration
11. **Expect** to land on `My Site` screen
   **Note:** Even though it could be an option to land on the screen from the web link (`Stats`), this wouldn't be ideal because users could open the Jetpack app after a long time.

Upon re-opening the app, I saw the migration error screen

#### 6. ✅ Jetpack App: Exit migration and login from `web link`

#### 7. ✅ Jetpack App: Open from `web link` without migration

#### 8. ❌ WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

1. 🆕 Clear WordPress app data (should be the case)
2. Open WordPress app and login
3. 🆕 Clear Jetpack app data
4. ☑️ Set **WordPress** app to handle verified links
5. ↗️ Open `Notifications` web link
6. **Expect** to see `Open links in Jetpack?` screen
7. ☑️ Set **Jetpack** app to handle verified links
8. Return to WordPress app
9. Tap `Open links in Jetpack`
10. **Expect** to see migration flow
15. Proceed to finish migration
16. ❌ **Expect** to land on `Notifications` screen

Everything worked about this except that the last step it went to my site screen.

#### 9. ✅ Regression: Jetpack App migration not eligible with Local Draft Content

</details>

I thought that perhaps the locale update could be the culprit (since this is the first time testing these flows with that feature enabled), so I tried testing without changes to the app settings in WordPress, but also got a failure (step 2 failed, and I didn't not test the remaining scenarios, since this disproved my hunch :disappointed: )

<details>
<summary>
These tests were performed on a Pixel 3a (physical device), without changes to WordPress app settings
</summary>

#### 1. ✅ Jetpack App: Complete migration flow from `web link`

#### 2. ❌ Jetpack App: Complete migration flow from `deep link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ↗️ Open `Jetpack Reader` deep link with Jetpack app (just once)
4. **Expect** to see migration flow
5. Proceed to finish migration
6. ❌ **Expect** to land on the `Reader` screen

I landed on the my site screen.

</details>

Of the scenarios that failed, #&zwj;5 seems like the most severe since it results in an error screen in the migration flow, while the others seem less severe (incomplete navigation to the deep link destination, which could be remedied by the user tapping the link once more).

I also double checked [my rebase](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17653-jp-migration-web-links--suggestion-rebase) by also creating [a new branch from this one](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...throwaway-issue-17653-merged-with-trunk) and performing the same conflict resolutions on a regular merge from `trunk`, and confirmed that the branches are identical with `git diff` to rule out errors made while rebasing / resolving conflicts. :man_shrugging:

ovitrif commented:
> Thanks for these updates Ovi 👍 .
> 
> Since conflicts have arisen, I created [a new branch with a rebase of this branch](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17653-jp-migration-web-links--suggestion-rebase) (which resolves the conflicts, and also splits out the `wip` commit from my earlier suggestion branch). However, after resolving the conflicts and retesting the above scenarios, I found that some of them are no longer working as expected 😞 . I haven't had a chance to dig into why yet, so hoping it is something simple 🤞 .

@mkevins 👋  Thank you for preparing a [better version history for these changes](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17653-jp-migration-web-links--suggestion-rebase) and for your effort to validate the PR over the latest updates on `trunk` 🙇 

> I thought that perhaps the locale update could be the culprit (since this is the first time testing these flows with that feature enabled), so I tried testing without changes to the app settings in WordPress, but also got a failure (step 2 failed, and I didn't not test the remaining scenarios, since this disproved my hunch 😞 )

Good thing we're retesting this together with [language migration](https://github.com/wordpress-mobile/WordPress-Android/pull/17768), as now even more things are coming together and it was expectable there might be a chance some of the fixes that were tested individually will need adjustments to integrate correctly with the others.

> Of the scenarios that failed, #‍5 seems like the most severe since it results in an error screen in the migration flow, while the others seem less severe (incomplete navigation to the deep link destination, which could be remedied by the user tapping the link once more).

I agree, 5️⃣ should definitely be improved!


I hope I'll have time this afternoon to re-test and investigate what causes the unexpected behavior 👍. Thanks a lot for the detailed reporting, it's definitely great to be able to narrow down the scope of the investigation already thanks to your work 🥇

ovitrif committed:
Fix deep linking to reader screen in JP app

ovitrif committed:
Replace also with let for consistency

ovitrif committed:
Reset blogging reminders sync flag isFirstTry flags

Since the blogging reminders migration were standardised, the prefs flag marking if syncing was attempted should be reset in order to enable resuming the flow after force-closing the app, just like it's done for all the other data migration flags.

ovitrif committed:
Merge branch 'trunk' into issue/17653-jp-migration-web-links

ovitrif commented:
@mkevins Thank you for all the help!

I've tested [your rebase branch](https://github.com/wordpress-mobile/WordPress-Android/compare/trunk...issue/17653-jp-migration-web-links--suggestion-rebase) and the same test cases failed for me as well.

After looking into what's working unexpected, I found a few incompatibilities with newer changes on `trunk` and proceeded to adjust that code. The following changes fixed the failing test cases for me:

- https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/99c5fa3d875bd04b2946bde698ce265527fee4c5 fixed the cases when instead of `Reader` or `Notifications` the app landed on `My Site`:
  - [x] ❌ Jetpack App: Complete migration flow from `deep link`
  - [x] ❌ Jetpack App: Interrupt and resume migration from `web link`
  - [x] ❌ WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

- https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/ef7467cc0db31fffabf74735e2ec0caf419c60d3 fixed the issue with resuming migration after force-closing the app
  - [x] ❌ Jetpack App: Force close migration from `web link` and restart

Both fixes address issues that I suspect to reproduce on `trunk` via the same scenarios from the test cases of this PR, so my guess is these are needed as well 😉 

I also added an intent of type `intent://home` (Jetpack Home) and validated it works 🚀

mkevins committed:
Revert "Replace also with let for consistency"

This reverts commit f8378e71752106f027bd0bb090b60d33f8a09226.

mkevins reviewed (approved):
> After looking into what's working unexpected, I found a few incompatibilities with newer changes on trunk and proceeded to adjust that code. The following changes fixed the failing test cases for me

Awesome!! Great work tracking down the source of the issues and restoring the behavior to it's working state! The code looks good, and all tests (including the most recent one with the intent link) are testing very well for me.

I tested the following on a Pixel 3a (physical device):

#### 1. ✅ Jetpack App: Complete migration flow from `web link`

#### 2. ✅ Jetpack App: Complete migration flow from `deep link`

#### 3. ✅ Jetpack App: Complete migration flow from `post link`

#### 4. ✅ Jetpack App: Interrupt and resume migration from `web link`

#### 5. ✅ Jetpack App: Force close migration from `web link` and restart

#### 6. ✅ Jetpack App: Exit migration and login from `web link`

1. **Prerequisite**: WordPress app installed and logged in (should be the case)
2. 🆕 Clear Jetpack app data
3. ☑️ Set Jetpack app to handle verified links (should be the case)
4. ↗️ Open `Notifications` web link
5. **Expect** to see migration flow
6. On the migration screen navigate back
7. **Expect** to see login screen
8. Login
9. **Expect** to land on the `My Site` screen
10. **Verify** that your data is not migrated (if you used the same account from WP app)

I landed on the notifications screen after login. I think this is a feature, though, and not a bug. :+1: 

#### 7. ✅ Jetpack App: Open from `web link` without migration

#### 8. ✅ WordPress App: Open `web link` in WordPress app then Jetpack via `Open links in Jetpack`

#### 9. ✅ Regression: Jetpack App migration not eligible with Local Draft Content

#### 10. ✅ Perform step 1, but with the Jetpack Home link instead


As a bonus, I also tested these scenarios with the WordPress app having a different "per-app" language setting and theme from the OS default, and everything works as expected with those conditions as well :tada: .

Note: I pushed one single revert (https://github.com/wordpress-mobile/WordPress-Android/pull/17742/commits/f23ac994c6bea55d5a9088af65929de14ddf5d9d), since `also` was intentionally used (instead of `let`). The reason is that this way, if the return type of `ActivityLauncher.openDeepLinkAfterJPMigration` is ever changed to an optional, it will not result evaluation of the RHS of the elvis operator (whereas with `let` it could, e.g. if the block evaluates to `null`).

Thanks for all your effort in writing the test scenarios, performing them, and working to find a solution to the issues after the merge!
